[Cache] Add (filesystem|phpfiles) cache (adapter|simple) prune method and prune command

robfrawley opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #21764, https://github.com/symfony/symfony/issues/21764#issuecomment-313425723
| License       | MIT
| Doc PR        | symfony/symfony-docs#8209

As requested in https://github.com/symfony/symfony/issues/21764#issuecomment-313425723, this PR adds a `prune()` method to [`FilesystemTrait`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Cache/Traits/FilesystemTrait.php). This placement seems reasonable as it exposes the method in [`FilesystemAdapter`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php) and [`FilesystemCache`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Cache/Simple/FilesystemCache.php).

The return value is a `bool` representing either a partial or complete failure (when `false`) *or* complete success (when `true`).

Once the API for the `prune` method is confirmed, I'll introduce a documentation PR, as well.

---

*Stale-detection implementation:* The file modification time is used to determine if a cache item should be pruned. This seems reasonable, given the use of [`touch` in the common trait](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Cache/Traits/FilesystemCommonTrait.php#L90). Interestingly, though, the [`doFetch` method](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Cache/Traits/FilesystemTrait.php#L38) uses the timestamp saved at the top of the file itself to determine the stale state. Should this latter implementation be used for `prune` as well (or is the current one ok), for example:

```php
foreach (new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($this->directory, \FilesystemIterator::SKIP_DOTS), \RecursiveIteratorIterator::LEAVES_ONLY, \RecursiveIteratorIterator::CATCH_GET_CHILD) as $file) {
    if ($h = @fopen($file, 'rb')) {
        if ($time >= (int) $expiresAt = fgets($h)) {
            fclose($h);
            if (isset($expiresAt[0])) {
                $okay = (@unlink($file) && !file_exists($file)) && $okay;
            }
        }
    }
}
```

robfrawley committed:
add (filesystem|phpfiles) cache (adapter|simple) prune method and prune command

- added `Symfony\Component\Cache\PruneableInterface` so PSR-6 or PSR-16 cache implementations can declare support
  for manual stale cache pruning
- added FilesystemTrait::prune() and PhpFilesTrait::prune() implementations
- now FilesystemAdapter, PhpFilesAdapter, FilesystemCache, and PhpFilesCache implement PruneableInterface and
  supports manual stale cache pruning
- Added `cache:pool:prune` command via `Symfony\Bundle\FrameworkBundle\Command\CachePoolPruneCommand` to allow
  manual stale cache item pruning of supported PSR-6 and PSR-16 cache pool implementations
- Added `Symfony\Bundle\FrameworkBundle\DependencyInjection\Compiler\CachePoolPrunerPass` compiler pass to fetch
  all cache pools implementing `PruneableInterface` and pass them to the command as an `IteratorArgument` so
  these references are lazy loaded by the command
- updated changelogs as appropriate

nicolas-grekas commented on src/Symfony/Component/Cache/Traits/FilesystemTrait.php:
```diff
@@ -23,6 +23,33 @@
+        $i = 0;
+
+        foreach (new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($this->directory, \FilesystemIterator::SKIP_DOTS), \RecursiveIteratorIterator::CHILD_FIRST, \RecursiveIteratorIterator::CATCH_GET_CHILD) as $file) {
+            // remove empty directories
```
Empty dirs shouldn't be removed: that creates a race condition.

robfrawley replied:
Noted. Will be removed in my next push.

nicolas-grekas commented on src/Symfony/Component/Cache/Traits/FilesystemTrait.php:
```diff
@@ -23,6 +23,33 @@
+     *
+     * @return int
+     */
+    public function prune($time = null)
```
Why allow time configuration?

robfrawley replied:
Figured it allows you to prune cache entries up to a specific time (if other than the current time). Someone may want to "removal all entries expiring in the next hour" or whatever. Do you not see any value in this?

nicolas-grekas commented on src/Symfony/Component/Cache/Traits/FilesystemTrait.php:
```diff
@@ -23,6 +23,33 @@
+            }
+        }
+
+        return $i;
```
A boolean would be enough IMHO

robfrawley replied:
I've gone ahead and changed the implementation to return `bool` for now; please advise if you find value in the above and I can revert; for now, it's as requested!

nicolas-grekas commented on src/Symfony/Component/Cache/Traits/FilesystemTrait.php:
```diff
@@ -23,6 +23,25 @@
+
+        foreach (new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($this->directory, \FilesystemIterator::SKIP_DOTS), \RecursiveIteratorIterator::LEAVES_ONLY, \RecursiveIteratorIterator::CATCH_GET_CHILD) as $file) {
+            if ($time >= $file->getMTime()) {
+                $okay = (@unlink($file) && !file_exists($file)) && $okay;
```
I think the first && should be an ||

nicolas-grekas reviewed (commented):
Okay doesn't exist, it's ok :) Maybe $pruned?
Also about the time arg, I personally don't see the need, so I'll let others confirm if this is really useful.

robfrawley commented:
@nicolas-grekas Just noticed https://github.com/symfony/symfony/issues/21764#issuecomment-284322178 where you mentioned adding a `PruneableInterface` to the cache component: is this something I should include in this PR?

nicolas-grekas commented:
@robfrawley yes please do, that would ease writting the command to prune (all prunable) pools.
But definitely the $time arg should not be in the interface to me.
If you could add the command into that PR, that'd be awesome.

robfrawley commented:
@nicolas-grekas What bundle should the command exist in?

nicolas-grekas commented:
FrameworkBundle since that's where the other cache:pool commands are I'd say.

robfrawley commented:
Re: the command. Do you want a complex implementation with a compiler pass that assigns all specially tagged services to a "global cache pruner" abstraction that is then consumed by the command (like the "cache clearer" implementation) or do you want a simple command that just consumes a list of cache pools, grabs 'em from the container, ensures they implement the prunable interface, and calls `prune()`?

robfrawley commented:
The latter would just look like:

```php
namespace Symfony\Bundle\FrameworkBundle\Command;

use Symfony\Component\Cache\PruneableInterface;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Component\Console\Style\SymfonyStyle;

/**
 * Cache pool pruner command.
 */
final class CachePoolPruneCommand extends ContainerAwareCommand
{
    /**
     * {@inheritdoc}
     */
    protected function configure()
    {
        $this
            ->setName('cache:pool:prune')
            ->addArgument('pools', InputArgument::IS_ARRAY | InputArgument::REQUIRED, 'A list of cache pools')
            ->setDescription('Prune cache pools')
            ->setHelp(<<<'EOF'
The <info>%command.name%</info> command prunes the given cache pools.

    %command.full_name% <cache pool 1> [...<cache pool N>]
EOF
            )
        ;
    }

    /**
     * {@inheritdoc}
     */
    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $io = new SymfonyStyle($input, $output);
        $container = $this->getContainer();

        foreach ($input->getArgument('pools') as $id) {
            if (!$container->has($id)) {
                throw new \InvalidArgumentException(sprintf('"%s" cache pool not found in container.', $id));
            }

            if (!($pool = $container->get($id)) instanceof PruneableInterface) {
                throw new \InvalidArgumentException(sprintf('"%s" is not a cache pool nor a cache clearer.', $id));
            }

            $io->comment(sprintf('Pruning cache pool: <info>%s</info>', $id));
            $pool->prune();
        }

        $io->success('Cache was successfully pruned.');
    }
}
```

nicolas-grekas commented:
See the cache pool clearer pass and command, it should be similar I guess

robfrawley commented:
@nicolas-grekas Is the most recent commit the direction you want this to go in? Just want to confirm before finishing up the implementation and ensuring all the required tests are present.

nicolas-grekas commented:
instead of a tag, what about constructing a service locator + array or ids (the pool services), that would be given to the command? I'm not sure I see the need for an intermediary pruner service, WDYT?

robfrawley commented:
That sounds reasonable. The PR implementation went in this direction as you mentioned mirroring how the "cache clear" functionality was implemented. Having said that, I absolutely agree that a simpler approach would be preferable and that an intermediary pruner service is unnecessary.

But, I'm not intimately familiar with service locators or their best practices. I get the basic concept, that you define a locator, specifying the services it should contain, and then use it like a `psr/container`. What I'm unsure of is the most appropriate way to dynamically ensure the locator is provided the required arguments (the purgeable pools). Should a compiler pass set the arguments for the service locator? And how does the compiler pass find the pools if not for tags? Is there another way to ensure the locator is passed all purgeable pools?

*Basically, how should the service locator be defined to allow dynamically passing all prunable pools to it?*

nicolas-grekas commented:
Do we need the command to be able to prune *one* specific pool given as argument? Or is pruning all prunable pools enough? I might thing *all* is enough, because pruning does remove things that are stale only.
If you agree, then we might use an iterator instead of a locator (`new IteratorArgument($listOfPools)`).
`$listOfPools` should be computed by looking at all pools in the compiler pass (via the cache.pool tag) and checking if they implement that interface.

robfrawley commented:
Give the latest changes a quick once-over. Some items still need to be cleaned up, but I think this is the direction you described. As for passing pools as arguments and using an argument iterator, I don't think those two have to be mutually exclusive. See the command implementation that optionally allows passing pool names. Thoughts? I don't have a strong opinion if you think we should remove the option to do so.

Also, right now the command filters the pools implementing the correct interface; I couldn't figure out how to filter pools during the compiler pass without reflection. If you think that's okay, the logic can be moved to the compiler pass by replacing `getCachePoolsIterator` with:

```php
/**
 * @param ContainerBuilder $container
 *
 * @return IteratorArgument
 */
private function getCachePoolsIterator(ContainerBuilder $container)
{
    $services = array_filter($this->getConcreteCachePoolServices($container), function ($name) use ($container) {
        $reflection = new \ReflectionClass($container->getDefinition($name)->getClass());

        return $reflection->implementsInterface(PruneableInterface::class);
    });

    return new IteratorArgument(array_combine($services, array_map(function ($id) {
        return new Reference($id);
    }, $services)));
}
```

RichardBradley commented on src/Symfony/Component/Cache/Tests/Simple/FilesystemCacheTest.php:
```diff
@@ -22,4 +22,52 @@ public function createSimpleCache($defaultLifetime = 0)
         return new FilesystemCache('', $defaultLifetime);
     }
+
+    public function testPrune()
```
Does this test fail without the other changes?
Looking at {{FilesystemCache::doContains}}, I think the cache will return false for expired entries, whether or not they have been pruned.

robfrawley replied:
This is a valid point; do you have any idea how to better test this functionality?

nicolas-grekas replied:
create a pool with only expired items, then check that prune did empty the dir?

robfrawley replied:
How about this approach: https://github.com/symfony/symfony/pull/23451/files#diff-6388c149445d15d020386824fc3440c5R63

RichardBradley commented on src/Symfony/Bundle/FrameworkBundle/Command/CachePoolPruneCommand.php:
```diff
@@ -0,0 +1,121 @@
+use Symfony\Component\DependencyInjection\Argument\RewindableGenerator;
+
+/**
+ * Cache pool pruner command.
```
I find "prune" a strange term for this. Is it in common use?

It seems to me that "garbage collection" or "gc" is a more commonly used term for deleting expired items.

RichardBradley replied:
There are lots of relevant Google hits for "cache prune", so maybe this is just me.

robfrawley replied:
Yeah; it's a common term. The `gc` term is common as well, but I think it is much less obvious as to its purpose (especially for lacking a CS background).

RichardBradley commented on src/Symfony/Component/Cache/Traits/FilesystemTrait.php:
```diff
@@ -23,6 +23,23 @@
+        $pruned = true;
+
+        foreach (new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($this->directory, \FilesystemIterator::SKIP_DOTS), \RecursiveIteratorIterator::LEAVES_ONLY, \RecursiveIteratorIterator::CATCH_GET_CHILD) as $file) {
+            if ($time >= $file->getMTime()) {
```
I think that `mtime` is not reliable enough for this use, especially on Windows or other "strange" filesystems that might be used for a file cache.

The current `FilesystemCache` implementation writes the expiry time to the first line of the file to avoid needing to use `mtime`, for this reason.

robfrawley replied:
I mentioned this in https://github.com/symfony/symfony/pull/23451#issue-241419054, none of the maintainers have mentioned `mtime` should not be used in favor of the in-file time. We'll see how @nicolas-grekas want to handle this.

nicolas-grekas replied:
checking the first line would be more robust indeed

robfrawley replied:
Done

RichardBradley commented on src/Symfony/Component/Cache/Traits/FilesystemTrait.php:
```diff
@@ -23,6 +23,23 @@
+        $time = time();
+        $pruned = true;
+
+        foreach (new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($this->directory, \FilesystemIterator::SKIP_DOTS), \RecursiveIteratorIterator::LEAVES_ONLY, \RecursiveIteratorIterator::CATCH_GET_CHILD) as $file) {
```
I think it's risky to duplicate the logic from `FileCache::getIterator` here. Can you not call that method directly?
Or at least put cross-referencing comments in the two locations? ("make sure to change x if you change here")

robfrawley replied:
I couldn't find any other instance of the following in any Symfony component...

```
new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($this->directory, \FilesystemIterator::SKIP_DOTS), \RecursiveIteratorIterator::LEAVES_ONLY, \RecursiveIteratorIterator::CATCH_GET_CHILD
```

nicolas-grekas replied:
we don't use CATCH_GET_CHILD in the current, code, we should align them, either adding it if it's missing, or removing it here in it's not really need.
In which case does it make a difference?

nicolas-grekas commented:
PhpFilesTrait could also benefit from pruning (should use unlink+opcache_invalidate)

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Command/CachePoolPruneCommand.php:
```diff
@@ -0,0 +1,121 @@
+    /**
+     * @param RewindableGenerator $pools
+     */
+    public function __construct(RewindableGenerator $pools)
```
`s/RewindableGenerator//` (iterable on master)

robfrawley replied:
Got it. What about https://github.com/symfony/symfony/pull/23451#issuecomment-314921923 --- should we use reflection in the compiler pass or a `instanceof` check in the command to filter the pools? And, should we get rid of the `pools` argument completely or not?

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/FrameworkBundle.php:
```diff
@@ -97,6 +98,7 @@ public function build(ContainerBuilder $container)
         $container->addCompilerPass(new LoggingTranslatorPass());
         $container->addCompilerPass(new AddCacheWarmerPass());
         $container->addCompilerPass(new AddCacheClearerPass());
+        $container->addCompilerPass(new CachePoolPrunerPass());
```
this pass should be run just before CachePoolClearerPass, so that Definition are resolved and you can check the class of each pool to check if it implements the new interface, and also to allow the removal of unused private cache pool (right now, the registration for the pruner prevents that).

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Command/CachePoolPruneCommand.php:
```diff
@@ -0,0 +1,121 @@
+    {
+        $pools = array();
+
+        foreach ($this->pools as $name => $pool) {
```
the issue of using an iterator is that this instantiate the skipped pools. 
so if we want to fine tune, using a service locator would prevent that (then we'd need an additional arg to the constructor: the list of pools found in the locator)

nicolas-grekas reviewed (commented):
shaping good :)

nicolas-grekas commented:
To check if a class implements some interface at compile time, you need to use `$container->getReflectionClass($class)`, see eg RegisterControllerArgumentLocatorsPass

robfrawley commented:
Still some work to do here; no need to review the latest commit. I'll push some additional changes by Wednesday, at which point this should be ready for a full review.

robfrawley commented:
Anyone have any idea why the `PruneableInterface` cannot be found on PHP `7.0` and `7.1` builds? See https://travis-ci.org/symfony/symfony/jobs/254946073#L2502. Interestingly, I am running both PHP versions on my local machine and the tests complete successfully locally...

nicolas-grekas commented:
> why the PruneableInterface cannot be found on PHP 7.0 and 7.1 

for deps=low line, look at the installed composer deps: Cache is 3.3, which cannot work since the interface is only in your PR. You need to update the lowest dep to 3.4 in fwb.
The other one installs master, which doesn't have the interface yet. It'll be green when the PR will be merged.

robfrawley commented:
Got it; thanks. I do have one concern about the prune implementations, but it relies on the answer to the following question: "Do we allow the use of different adapters at the same root path?" For example, do we support instantiating a `PhpFilesAdapter` and `FilesystemAdapter` with the same `directory` value (take, for example, `/tmp/cache-root`).

If so, the `prune` logic needs to check if the file found meets the format requirements of the specific adapter, otherwise reading the expiration will fail ([here](https://github.com/symfony/symfony/pull/23451/files#diff-9f5aec4225d61147061d9617681a4e74R35) and [here](https://github.com/symfony/symfony/pull/23451/files#diff-44745b984bf9578ef8f00218a3eadd05R46)) when a file from a different adapter is encountered.

Otherwise, I think I've addressed the majority of concerns/questions/etc above. One outlier is the prune command itself, which I'm not entirely positive as to the best implementation. I guess it would be way simpler if we just prune all pools and don't allow specifying specific pool service ids... thoughts?

nicolas-grekas commented:
let's prune all pools :)
> Do we allow the use of different adapters at the same root path

that's unsupported, each pool must have a separate namespace, ie the type of adapter should be considered as an implicit namespace

robfrawley commented:
Before your most-recent reply, I actually implemented a version of this pull request with a service locator that still allowed for passing specific pools to the command (see https://github.com/robfrawley/symfony/commit/301ee4a83b7bd30098e99d7a6decebd39f9beba0). It added a fair amount of complexity.

I don't have any strong feeling about which is best, so unless you say otherwise, we'll go with simply pruning all pools, as seen in the most-recent commit here.

This is one of the larger PRs I've contributed to Symfony, but I *think* this is ready to be reviewed, at your convenience.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPrunerPass.php:
```diff
@@ -0,0 +1,127 @@
+        $definition = $container->getDefinition($id);
+        $class = $definition->getClass();
+
+        while (!$class && $definition instanceof ChildDefinition) {
```
this logic is not required, we should hook after child def are resolved (see next comment)

robfrawley replied:
Without this logic, [this ChildDefinition test of the compiler pass](https://github.com/symfony/symfony/pull/23451/files#diff-ac294e6b38a782f221d2e7fc2c56e50fR45) fails. Due to the below comment, this will work properly in production and the test should be removed, as well, correct?

nicolas-grekas replied:
but this test is fictional so it should be removed :)

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/FrameworkBundle.php:
```diff
@@ -96,6 +97,7 @@ public function build(ContainerBuilder $container)
         $this->addCompilerPassIfExists($container, TranslatorPass::class);
         $container->addCompilerPass(new LoggingTranslatorPass());
         $container->addCompilerPass(new AddCacheWarmerPass());
+        $container->addCompilerPass(new CachePoolPrunerPass());
```
this pass must be registered next to CachePoolClearerPass, as TYPE_AFTER_REMOVING
this is to allow removable of unused (prunebale) cache pool, which the current code actually prevents, by always creating a reference to them for the command

robfrawley replied:
I follow everything short of "next to CachePoolClearerPass" --- does "next to" mean "before" or "after" CachePoolClearerPass?

nicolas-grekas commented on src/Symfony/Component/Cache/Tests/Adapter/FilesystemAdapterTest.php:
```diff
@@ -49,4 +49,62 @@ public static function rmdir($dir)
         rmdir($dir);
     }
+
+    public function testPrune()
```
can we move this to the parent test case (with a skip if the pool doesn't implement the interface)
same for PSR16?
that would require each $isHit closure to be turned into method on each specific pool's case (which I wouldn't call "ishit" btw, but maybe isPruned?)

robfrawley replied:
The four test cases don't share a common parent class within this component's scope. Should I create a trait that can be used by `AdapterTestCase` and `CacheTestCase`? Or perhaps a new `AbstractCacheTestCase` that the aforementioned classes can inherit from? Or should the logic be duplicated between the `Adapter` and `Simple` top-level test classes?

nicolas-grekas replied:
let's go with duplication between Adapter and Simple, which may allow both to diverge (eg populating the pool can then be done using each's interfaces)

nicolas-grekas commented on src/Symfony/Component/Cache/Traits/PhpFilesTrait.php:
```diff
@@ -82,7 +111,7 @@ protected function doSave(array $values, $lifetime)
         $ok = true;
         $data = array($lifetime ? time() + $lifetime : PHP_INT_MAX, '');
-        $allowCompile = 'cli' !== PHP_SAPI || ini_get('opcache.enable_cli');
+        $allowCompile = static::opcacheAllowCompile();
```
lets' keep this inline, that's a onliner

nicolas-grekas commented on src/Symfony/Component/Cache/Traits/PhpFilesTrait.php:
```diff
@@ -32,6 +32,35 @@ public static function isSupported()
+
+        set_error_handler($this->includeHandler);
+        try {
+            foreach (new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($this->directory, \FilesystemIterator::SKIP_DOTS), \RecursiveIteratorIterator::LEAVES_ONLY, \RecursiveIteratorIterator::CATCH_GET_CHILD) as $file) {
```
same comment about CATCH_GET_CHILD

robfrawley replied:
The `CATCH_GET_CHILD` flag causes iterator children calls to not throw an exception (for example if an unreadable directory is encountered) which made sense to me in the context of `prune`. Do you think it's needed here? Is it needed elsewhere? Or should be removed?

nicolas-grekas replied:
you're sending back my questions above :)
if I can decide: it's not used right now + nobody complained => let's continue to not use it at all

robfrawley replied:
Ha, yeah, I guess I did: I guess there _shouldn't_ be unreachable children in the cache root; I'll remove it. I'll push changes shortly that address all your comments.

robfrawley commented:
Ok; comments addressed. Let me know if there are any additional changes required, or if I should squash the commits or anything else I can contribute. 

Also, is the [`$container->getDefinition($id)->isAbstract()` check](https://github.com/symfony/symfony/pull/23451/files#diff-b66c2265ca4c725133c3d2077e8670a4R57) still required now, with the updated location of the compiler pass and the amended `PassConfig::TYPE_AFTER_REMOVING` type passed?

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPrunerPass.php:
```diff
@@ -0,0 +1,87 @@
+        $services = array();
+
+        foreach ($container->findTaggedServiceIds($this->cachePoolTag) as $id => $tags) {
+            if ($container->getDefinition($id)->isAbstract()) {
```
can be removed indeed

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPrunerPass.php:
```diff
@@ -0,0 +1,87 @@
+                continue;
+            }
+
+            $services[] = $id;
```
since this is internal, it should directly build the right structure: an id-to-ref map (and the array_combine above be remove)
also, I don't feel the need do keep this function, now that the caller is almost empty now. In should be merged with "process"

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPrunerPass.php:
```diff
@@ -0,0 +1,87 @@
+     *
+     * @return \ReflectionClass
+     */
+    private function resolveServiceReflectionClass(ContainerBuilder $container, $id)
```
same here, not sure about the need to keep a separate method we do that inline most of the time in existing passes

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPrunerPass.php:
```diff
@@ -0,0 +1,87 @@
+    {
+        $class = $container->getParameterBag()->resolveValue($container->getDefinition($id)->getClass());
+
+        if (null === $reflection = $container->getReflectionClass($class)) {
```
`if (!$reflection`, no need to be strict on the type here

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/FrameworkBundle.php:
```diff
@@ -97,6 +98,7 @@ public function build(ContainerBuilder $container)
         $container->addCompilerPass(new LoggingTranslatorPass());
         $container->addCompilerPass(new AddCacheWarmerPass());
         $container->addCompilerPass(new AddCacheClearerPass());
+        $container->addCompilerPass(new CachePoolPrunerPass(), PassConfig::TYPE_AFTER_REMOVING);
```
should be moved after CachePoolClearerPass for clarity

nicolas-grekas commented on src/Symfony/Component/Cache/PruneableInterface.php:
```diff
@@ -0,0 +1,23 @@
+namespace Symfony\Component\Cache;
+
+/**
+ * Interface for adapters and simple cache implementations that allow pruning entries.
```
pruning expired items

nicolas-grekas commented on src/Symfony/Component/Cache/Tests/Simple/CacheTestCase.php:
```diff
@@ -59,6 +69,57 @@ public function testNotUnserializable()
+            return $cache->has($name) && file_exists($getFileMethod->invoke($cache, $name));
+        };
+
+        $doSet = function ($name, $value, \DateInterval $expiresAfter = null) use ($cache) {
```
this closure is now trivial, it's not useful anymore

nicolas-grekas commented on src/Symfony/Component/Cache/Tests/Adapter/AdapterTestCase.php:
```diff
@@ -67,6 +72,64 @@ public function testNotUnserializable()
+
+        $cache = $this->createCachePool();
+
+        $isPruned = function ($name) use ($cache) {
```
this should be turned into a method on the concrete test case: the implementation is very specific to a filesystem cache and wouldn't work with any other kind of pruneable pool (there will be I'm sure :) )

nicolas-grekas commented on src/Symfony/Component/Cache/Tests/Simple/CacheTestCase.php:
```diff
@@ -59,6 +69,57 @@ public function testNotUnserializable()
+
+        $cache = $this->createSimpleCache();
+
+        $isPruned = function ($name) use ($cache) {
```
sam here: should be a method on concrete class

nicolas-grekas commented on src/Symfony/Component/Cache/Traits/FilesystemTrait.php:
```diff
@@ -23,6 +23,26 @@
+
+        foreach (new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($this->directory, \FilesystemIterator::SKIP_DOTS), \RecursiveIteratorIterator::LEAVES_ONLY) as $file) {
+            if ($h = @fopen($file, 'rb')) {
+                if ($time >= (int) $expiresAt = fgets($h)) {
```
to prevent nesting of two "if", use "if (!$h ...) => continue"

nicolas-grekas commented on src/Symfony/Component/Cache/Traits/FilesystemTrait.php:
```diff
@@ -23,6 +23,26 @@
+        foreach (new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($this->directory, \FilesystemIterator::SKIP_DOTS), \RecursiveIteratorIterator::LEAVES_ONLY) as $file) {
+            if ($h = @fopen($file, 'rb')) {
+                if ($time >= (int) $expiresAt = fgets($h)) {
+                    fclose($h);
```
the fclose should be outside the if block - at least the file needs to be closed in the "else" path

nicolas-grekas commented on src/Symfony/Component/Cache/Traits/FilesystemTrait.php:
```diff
@@ -23,6 +23,26 @@
+            if ($h = @fopen($file, 'rb')) {
+                if ($time >= (int) $expiresAt = fgets($h)) {
+                    fclose($h);
+                    $pruned = (isset($expiresAt[0]) && @unlink($file) && !file_exists($file)) && $pruned;
```
extra brackets to be removed

nicolas-grekas commented on src/Symfony/Component/Cache/Traits/PhpFilesTrait.php:
```diff
@@ -32,6 +32,35 @@ public static function isSupported()
+                list($expiresAt) = include $file;
+
+                if ($time >= $expiresAt) {
+                    $pruned = (@unlink($file) && !file_exists($file)) && $pruned;
```
extra brackets

nicolas-grekas reviewed (commented):
almost there :)

robfrawley commented:
@nicolas-grekas Ok; I think I properly addressed your comments. One additional question, now that `isPruned($cache, $name)` has been moved to the concrete test classes, should the `testPrune()` methods check for the existance of the `isPruned()` method and skip if it doesn't exist, [here](https://github.com/symfony/symfony/pull/23451/files#diff-e1e3d11734b8bdf5026e2b467821639aR78).

nicolas-grekas commented:
> check for the existance of the isPruned() method and skip if it doesn't exist

I think it should not skip, a hard fail would be the best in fact, to force an pool author to implement this method for pruneable pools

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/cache.xml:
```diff
@@ -100,6 +100,11 @@
             </call>
         </service>
 
+        <service id="cache.command.pool_pruner" class="Symfony\Bundle\FrameworkBundle\Command\CachePoolPruneCommand" public="true">
```
does the command need to be public really?

robfrawley replied:
I don't think so... removed `public` attribute.

nicolas-grekas commented on src/Symfony/Component/Cache/Tests/Adapter/AdapterTestCase.php:
```diff
@@ -67,6 +72,59 @@ public function testNotUnserializable()
+        $doSet('qux', 'qux-val', new \DateInterval('PT80S'));
+
+        $cache->prune();
+        $this->assertTrue($this->isPruned($cache, 'foo'));
```
to be consistent with the naming, all true/false assertions should be swapped, and isPruned should return true when the key is pruned (it does the opposite right now isn't it?)

robfrawley replied:
You're right; that is due to how it used to be named `isHit` --- I'll reverse the return now that the name has changed. ;-)

robfrawley replied:
Do you have a preference between
```php
return !$cache->getItem($name)->isHit() && !file_exists($getFileMethod->invoke($cache, $name))
```
and
```php
return !($cache->getItem($name)->isHit() && file_exists($getFileMethod->invoke($cache, $name)))
```

nicolas-grekas replied:
that's not the same :) 
what about just `return !file_exists($getFileMethod->invoke($cache, $name);`?
because basically that's what the "prune" method does: clean files

robfrawley replied:
S*it, you're right. :-( I must be getting tired... I think just checking the file is a good call.

nicolas-grekas commented on src/Symfony/Component/Cache/Tests/Simple/CacheTestCase.php:
```diff
@@ -59,6 +69,48 @@ public function testNotUnserializable()
+        $cache->set('qux', 'qux-val', new \DateInterval('PT80S'));
+
+        $cache->prune();
+        $this->assertTrue($this->isPruned($cache, 'foo'));
```
same here, true/false should be swapped

nicolas-grekas commented on src/Symfony/Component/Cache/Traits/FilesystemTrait.php:
```diff
@@ -23,6 +23,29 @@
+                $pruned = isset($expiresAt[0]) && @unlink($file) && !file_exists($file) && $pruned;
+            }
+
+            fclose($h);
```
that won't work on Windows, where you cannot unlink a file when all handlers to it are not closed

robfrawley replied:
F*** Windows. :-) But really, I guess just add an `else` branch?

```php
if ($time >= (int) $expiresAt = fgets($h)) {
    fclose($h);
    $pruned = isset($expiresAt[0]) && @unlink($file) && !file_exists($file) && $pruned;
} else {
    fclose($h);
}
```

nicolas-grekas replied:
lgtm

robfrawley commented:
Additional changes pushed. If it's an okay time to squash the commits made for your comments into the original commit, let me know (don't want to do so before you can confirm the changes, as it would be harder to do so when squashed, obviously). Thanks for taking the time to so thoroughly review this PR!

Not for this PR, but more generally out of interest (and possibly submission of additional PRs), you've mentioned that other PSR-(6|16) implementations may benefit from implementing `PruneableInterface`: which ones come to mind for you as benefitting most?

nicolas-grekas reviewed (approved):
thanks for the work!

nicolas-grekas commented:
> which ones come to mind for you as benefitting most?

the PdoAdapter first, then maybe ChainAdapter could propagate, and last to my mind comes the ApcuAdapter, looking for items that have an old `$version.'@'.$namespace` (but not sure about it esp. since this will be unusable from the command because of different shared memory segments used by the webserver/FPM)

robfrawley commented:
@nicolas-grekas I know #23603 is still fluid and unreviewed, but for the purpose of the documentation PR I am working on symfony/symfony-docs#8209, everything in **this** PR should be fairly solid at this point, right?

nicolas-grekas commented:
Thank you @robfrawley.

Toflar commented:
@nicolas-grekas there's some design issue here when using the `TagAwareAdapter` because you cannot access the inner `$itemsAdapter`, you cannot ask for `implements PruneableInterface` in such a case:

```php
$cache = new TagAwareAdapter(new FilesystemAdapter());
// Does $cache implement Pruneable now?
```

I figured the easiest way would be to have a `PruneableTagAwareAdapter` like so:

```php
class PruneableTagAwareAdapter extends TagAwareAdapter implements PruneableInterface
{
    /**
     * @return bool
     */
    public function prune()
    {
        return $this->itemsAdapter->prune();
    }
}
```

However, the `$itemsAdapter` property is `private` so I would need to reimplement everything.
Can we improve this by either making the properties of `TagAwareAdapter` `protected` or have it implement `PruneableInterface` and check itself on `$this->itemsAdapter` again? Wdyt?

robfrawley commented:
@Toflar Looks like your concern is reasonable, but it is generally better to open a new issue/pr than to comment on a closed one, as it removes visibility of the issue from anyone not directly involved in the existing pr, and not everyone actually follows-up on comments to closed issues/prs. :-)

That said, why not just make the adapter prune-aware instead of adding complexity with an additional interface:

```php
class TagAwareAdapter implements PruneableInterface
{
    /**
     * @return bool
     */
    public function prune()
    {
        if ($this->itemsAdapter instanceof PruneableInterface) {
            return $this->itemsAdapter->prune();
        }

        return false;
    }
}
```

Toflar commented:
Yeah, that's what I meant with my last sentence :) Will create a PR directly then :)

kbond commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPrunerPass.php:
```diff
@@ -0,0 +1,60 @@
+    private $cacheCommandServiceId;
+    private $cachePoolTag;
+
+    public function __construct($cacheCommandServiceId = 'cache.command.pool_pruner', $cachePoolTag = 'cache.pool')
```
Pretty sure this is broken right now as `cache.command.pool_pruner` doesn't exist.
